File: leetcode/lru_cache.py
#!/usr/bin/env python3

from collections import OrderedDict
import sys
import unittest

# number: 146
# title: LRU Cache
# url: https://leetcode.com/problems/lru-cache/
# section: linked list
# difficulty: medium
# tags: hash table, linked list, design, doubly-linked list, top 150, meta, optiver, citadel, grind 75

# constraints
# 1 <= capacity <= 3000
# 0 <= key <= 10^4
# 0 <= value <= 10^5
# At most 2 * 10^5 calls will be made to get and put.

# solution: OrderedDict
# complexity
# run-time: O(1) for both put/get
# space: O(capacity)
class LRUCache:

    def __init__(self, capacity: int):
        self.__cache = OrderedDict()

        if capacity <= 0:
            raise ValueError("Invalid capacity")

        self.__capacity = capacity

    def get(self, key: int) -> int:
        #print(f"get cache:{self.__cache} for key:{key}")

        if key not in self.__cache:
            return -1

        # mark as recent by moving to end of queue
        self.__cache.move_to_end(key)

        return self.__cache[key]

    def put(self, key: int, value: int) -> None:
        #print(f"put cache:{self.__cache} for key:{key}")

        # overwrite existing key
        if key in self.__cache:
            self.__cache[key] = value
            # end of queue: like deque.append()
            self.__cache.move_to_end(key)
            return

        # at capacity, free up
        if len(self.__cache) == self.__capacity:
            # from beginning of queue: like deque.popleft()
            self.__cache.popitem(last=False)

        # new key
        self.__cache[key] = value

# solution: dict (>= 3.7)
# complexity
# run-time: O(1) for both put/get
# space: O(capacity)
class LRUCache2:

    def __init__(self, capacity: int):
        self.__cache = {}

        if capacity <= 0:
            raise ValueError("Invalid capacity")

        self.__capacity = capacity

    def get(self, key: int) -> int:
        #print(f"get cache:{self.__cache} for key:{key}")

        if key not in self.__cache:
            return -1

        # mark as recent by readding
        val = self.__cache[key]
        del self.__cache[key]
        self.__cache[key] = val

        return self.__cache[key]

    def put(self, key: int, value: int) -> None:
        #print(f"put cache:{self.__cache} for key:{key}")

        # mark existing key as recent by readding
        if key in self.__cache:
            del self.__cache[key]
        # at capacity, free up LRU
        elif len(self.__cache) == self.__capacity:
            old_key = next(iter(self.__cache))
            del self.__cache[old_key]

        # add/update key
        self.__cache[key] = value

class TestLRUCache(unittest.TestCase):

    def test_0_capacity(self):
        capacity = 0
        self.assertRaises(ValueError, LRUCache, capacity)
        self.assertRaises(ValueError, LRUCache2, capacity)

    def test_neg_capacity(self):
        capacity = -5
        self.assertRaises(ValueError, LRUCache, capacity)
        self.assertRaises(ValueError, LRUCache2, capacity)

    def test_lrucache(self):
        capacity = 2
        obj = LRUCache(capacity)
        obj.put(1, 1); # cache is {1=1}
        obj.put(2, 2); # cache is {1=1, 2=2}
        self.assertEqual(obj.get(1), 1)
        obj.put(3, 3); # LRU key was 2, evicts key 2, cache is {1=1, 3=3}
        self.assertEqual(obj.get(2), -1)
        obj.put(4, 4); # LRU key was 1, evicts key 1, cache is {4=4, 3=3}
        self.assertEqual(obj.get(1), -1)
        self.assertEqual(obj.get(3), 3)
        self.assertEqual(obj.get(4), 4)

    def test_lrucache2(self):
        capacity = 2
        obj = LRUCache2(capacity)
        obj.put(1, 1); # cache is {1=1}
        obj.put(2, 2); # cache is {1=1, 2=2}
        self.assertEqual(obj.get(1), 1)
        obj.put(3, 3); # LRU key was 2, evicts key 2, cache is {1=1, 3=3}
        self.assertEqual(obj.get(2), -1)
        obj.put(4, 4); # LRU key was 1, evicts key 1, cache is {4=4, 3=3}
        self.assertEqual(obj.get(1), -1)
        self.assertEqual(obj.get(3), 3)
        self.assertEqual(obj.get(4), 4)

if __name__ == '__main__':
    sys.exit(unittest.main())

File: leetcode/lru_cache.cpp
#include <list>
#include <stdexcept>
#include <tuple>
#include <unordered_map>

// number: 146
// title: LRU Cache
// url: https://leetcode.com/problems/lru-cache/
// section: linked list
// difficulty: medium
// tags: hash table, linked list, design, doubly-linked list, top 150, meta,
// optiver, citadel

// constraints
// 1 <= capacity <= 3000
// 0 <= key <= 10^4
// 0 <= value <= 10^5
// At most 2 * 10^5 calls will be made to get and put.

// solution: unordered_map + list
// complexity
// run-time: O(1) for both put/get
// space: O(capacity)
#include <list>
#include <stdexcept>
#include <tuple>
#include <unordered_map>

class LRUCache {
public:
    LRUCache(int capacity)
    : d_capacity(capacity) {
        if (capacity <= 0) {
            throw std::invalid_argument("invalid capacity"); 
        }
    }

    int get(int key) {
        // refresh: make recently used if found
        return update(key, -1, false);
    }

    void put(int key, int value) {
        // update existing key if found
        if (update(key, value) != -1) {
            return;
        }

        // at capacity
        if (d_cache.size() == d_capacity) {
            int lruKey = std::get<0>(d_list.front());
            d_cache.erase(lruKey);
            d_list.pop_front();
        }

        // add new key
        d_list.push_back(std::make_tuple(key, value));

        // save an iterator to it
        d_cache[key] = std::prev(d_list.end());
    }

private:
    // refersh: make recently used if found & update
    int update(int key, int value, bool overwrite=true) {
        KeyListMap::iterator cacheIter = d_cache.find(key);
        if (cacheIter == d_cache.end()) {
            return -1;
        }

        ListTuples::iterator listIter = cacheIter->second;
        if (!overwrite) {
            value = std::get<1>(*listIter);
        }

        d_list.erase(listIter);
        d_list.push_back(std::tuple<int, int>(key, value));

        // get iter to last element, not end placeholder
        d_cache[key] = std::prev(d_list.end());

        return value;
    }

    // key, value tuples
    using ListTuples = std::list<std::tuple<int, int>>;
    // key: List iterator
    using KeyListMap = std::unordered_map<int, ListTuples::iterator>;
    KeyListMap d_cache;
    ListTuples d_list;
    int d_capacity;
};

// TODO: add unit tests

// Your LRUCache object will be instantiated and called as such:
int capacity = 2;
int key = 1;
LRUCache obj = LRUCache(capacity);
int param_1 = obj.get(key);
obj.put(key,value);